File: c++/1721-swapping-nodes-in-a-linked-list.cpp
class Solution {
public:
    ListNode* swapNodes(ListNode* head, int k) {
        if(head == NULL || head -> next == NULL) 
            return head;

        ListNode *ptr = head, *beg = head, *end = head;
        int a = 0;
        while(ptr != NULL) {
            a++;
            if(a == k) 
                beg = ptr;
            if(a >= k + 1) 
                end = end -> next;
            ptr = ptr -> next;
        }
        int temp = beg -> val;
        beg -> val = end -> val;
        end -> val = temp;
        return head;
    }
};

File: python/1700-1799/1721-swapping-nodes-in-a-linked-list.py
class Solution:
    def swapNodes(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        fast = slow = head
        for _ in range(k - 1):
            fast = fast.next
        p = fast
        while fast.next:
            fast, slow = fast.next, slow.next
        q = slow
        p.val, q.val = q.val, p.val
        return head
